recursion.md

```diff
@@ -0,0 +1,172 @@
+# Mental models for recursion
+
+## Table of contents
+
+### [Introduction to recursion](#introduction-to-recurision-1)
+### [Exponent](#exponent-1)
+
+## Introduction to recursion
+
+Problems that you can solve with recursion have a recursive structure. This
+usually means that the problem, a) is possible to reduce into smaller problems,
+and b) has a knowable base case.
+
+Tips for writing recursive problems:
+
+1) Use simple base cases
+
+Avoiding "arms length" base cases can be a good idea to keep the logic of your
+code simple and easy to understand. For example, if you are writing a function
+that can compute raise a base to a power of `exp`, base cases like this are
+unnecessarily specific, and take away from some of the simplicity you can
+achieve in a recursive solution.
+
+```rb
+return 1 if exp == 0
+return base if exp == 1
+return base * base if exp == 2
+```
+
+If you structure your problem right, all you need for your base case is the
+first one - `return 1 if exp == 0`.
+
+2) Be clear and consistent in terms of what's the return value of the recursive algorithm.
+
+## Exponent
+
+Write a function that takes a number (base) and an exponent. Raise the base to
+the power of a given exponent.
+
+**Solution 1: Simplest recursive solution - positive numbers**
+
+This is perhaps the most straight-forward way to solve the exponent/to the power
+of problem using recursion. Note that this first solution will only work for
+positive exponents.
+
+```rb
+def power(base, exponent)
+  if exponent == 0
+    return 1
+  else
+    power(base, exponent - 1)
+  end
+end
+```
+
+Here's an example of what the callstack would look if we call `power(2,3)`
+
+exponent|recursive call               |return value
+--------|-----------------------------|------------
+0       |                             | 1
+1       |2 * power(2, (1-1)) => 2 * 1 | 2
+1       |2 * power(2, (2-1)) => 2 * 2 | 4
+2       |2 * power(2, (2-1)) => 2 * 4 | 8
+
+Runtime - O(n):
+
+Essentially, what we're trying to find the cost of exponent ``(exponent - 1)`` + the base case`( 0(1))`. This means that our runtime complexity is `O(exponent)`.
```

Comment:
I'm not 100% certain why the runtime complexity is 0(n). I feel like we're making n-1 calls every n elements in the size of the problem.

Reply:
From the time the program is executed to the time it ends, there are only ever two possibilities for the program to encounter: either it hit its base case or it didn't.

Every operation other than the recursive call is an O(1) operation. Therefore, the time complexity of the method is O(number of recursive calls + 1) which can be simplified to O(number of recursive calls). Since the method makes an *exponent* number of recursive calls, the runtime of this algorithm is O(exponent)